src/Uno.Toolkit.UI/Controls/ZoomContentControl/ZoomContentControl.cs

```diff
@@ -0,0 +1,467 @@
+		{
+			return false;
+		}
+
+		var canMove = false;
+		canMove |= CanScrollLeft() && _movementDirection.Horizontal is true;
+		canMove |= CanScrollRight() && _movementDirection.Horizontal is false;
+		canMove |= CanScrollUp() && _movementDirection.Vertical is true;
+		canMove |= CanScrollDown() && _movementDirection.Vertical is false;
+
+		return canMove;
+	}
+
+	private bool CanScrollUp() => !ContentBoundsVisibility.HasFlag(BoundsVisibilityFlag.Top);
+	private bool CanScrollDown() => !ContentBoundsVisibility.HasFlag(BoundsVisibilityFlag.Bottom);
+	private bool CanScrollLeft() => !ContentBoundsVisibility.HasFlag(BoundsVisibilityFlag.Left);
+	private bool CanScrollRight() => !ContentBoundsVisibility.HasFlag(BoundsVisibilityFlag.Right);
+
+	private async void UpdateHorizontalScrollBarValue()
+	{
+		HorizontalScrollValue = -1 * HorizontalOffset;
+		await RaiseRenderedContentUpdated();
+	}
+
+	private void IsActiveChanged()
+	{
+		if (ResetZoomAndOffsetWhenInactive && !IsActive)
+		{
+			RemoveOffset();
+			ResetZoom();
+		}
+		if (_scrollH is not null)
+		{
+			_scrollH.Visibility = IsActive ? Visibility.Visible : Visibility.Collapsed;
+		}
+		if (_scrollV is not null)
+		{
+			_scrollV.Visibility = IsActive ? Visibility.Visible : Visibility.Collapsed;
+		}
+	}
+
+	private void UpdateScrollLimits()
+	{
+		if (_presenter?.Content is FrameworkElement fe)
+		{
+			var verticalScroll = Math.Max(0, (fe.ActualHeight * ZoomLevel) - ViewportHeight);
+			var horizontalScroll = Math.Max(0, (fe.ActualWidth * ZoomLevel) - ViewportWidth);
+
+			HorizontalMaxScroll = horizontalScroll / 2;
+			VerticalMaxScroll = verticalScroll / 2;
+
+			HorizontalMinScroll = -1 * HorizontalMaxScroll;
+			VerticalMinScroll = -1 * VerticalMaxScroll;
+		}
+	}
+
+	private void CoerceZoomLevel()
+	{
+		ZoomLevel = Math.Clamp(ZoomLevel, MinZoomLevel, MaxZoomLevel);
+	}
+
+	// Template handling
+	protected override void OnApplyTemplate()
+	{
+		T FindTemplatePart<T>(string name) where T : class =>
+			(GetTemplateChild(name) ?? throw new Exception($"Expected template part not found: {name}"))
+			as T ?? throw new Exception($"Expected template part '{name}' to be of type: {typeof(T)}");
+
+		_presenter = FindTemplatePart<ContentPresenter>(TemplateParts.Presenter);
+		_scrollV = FindTemplatePart<ScrollBar>(TemplateParts.VerticalScrollBar);
+		_scrollH = FindTemplatePart<ScrollBar>(TemplateParts.HorizontalScrollBar);
+
+		ResetOffset();
+		ResetZoom();
+
+		RegisterToControlEvents();
+		RegisterPointerHandlers();
+	}
+
+	// Event handlers
+	private void OnLoaded(object sender, RoutedEventArgs e)
+	{
+		CenterContent();
+	}
+
+	private void OnSizeChanged(object sender, SizeChangedEventArgs args)
+	{
+		UpdateContentBoundsVisibility();
+		if (IsLoaded && AutoZoomToCanvasOnSizeChanged)
+		{
+			FitToCanvas();
+		}
+	}
+
+	private void RegisterToControlEvents()
+	{
+		if (_presenter?.Content is FrameworkElement { } fe)
+		{
+			fe.LayoutUpdated += (s, e) =>
```

Comment:
I'd be super super super careful about using `LayoutUpdated`. It's in many cases not really the right thing to do. You should treat this event as a static event that's not actually associated with a specific element.